pyccel/codegen/printing/ccode.py

```diff
@@ -665,31 +665,52 @@ def _print_IndexedElement(self, expr):
-                                isinstance(end.args[0], LiteralInteger):
-                            end = PyccelMinus(base_shape[i], end.args[0])
-                        if ind.start is None:
-                            start = 0
-                        if ind.end is None:
-                            end = base.shape[i]
-                        inds[i] = Slice(start, end)
-                    else:
-                        #setting the Slice start and end to their correct value when try to get a view with scalar index
-                        inds[i] = Slice(ind, PyccelAdd(ind, LiteralInteger(1)))
+                        inds[i] = self._slice_elements_managements(ind, PyccelArraySize(base, i),
+                            allow_negative_indexes)
                 inds = [self._print(i) for i in inds]
                 return "array_slicing(%s, %s)" % (base_name, ", ".join(inds))
             inds = [self._print(i) for i in inds]
         else:
             raise NotImplementedError(expr)
         return "%s.%s[get_index(%s, %s)]" % (base_name, dtype, base_name, ", ".join(inds))
 
+    def _slice_elements_managements(self, _slice, shape, allow_negative_index):
+
+        start = LiteralInteger(0) if _slice.start is None else _slice.start
+        end = shape if _slice.end is None else _slice.end
+
+        # negative start and end in slice
+        if isinstance(start, PyccelUnarySub) and isinstance(start.args[0], LiteralInteger):
+            start = PyccelMinus(shape, start.args[0])
+        elif allow_negative_index and not isinstance(start, LiteralInteger):
+            start = IfTernaryOperator(PyccelLt(start, LiteralInteger(0)),
+                            PyccelMinus(shape, start), start)
+
+        if isinstance(end, PyccelUnarySub) and isinstance(end.args[0], LiteralInteger):
+            end = PyccelMinus(shape, end.args[0])
+        elif allow_negative_index and not isinstance(end, LiteralInteger):
+            end = IfTernaryOperator(PyccelLt(end, LiteralInteger(0)),
+                            PyccelMinus(shape, end), end)
+
+        # steps in slices
+        step = _slice.step
+        if step is not None :
+            # negative step in slice
+            if isinstance(step, PyccelUnarySub) and isinstance(step.args[0], LiteralInteger):
+                start = shape if _slice.start is None else start
+                end = LiteralInteger(0) if _slice.end is None else end
+
+            # variable step in slice
+            elif allow_negative_index and not isinstance(step, LiteralInteger):
+                start = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)), start, end)
+                end = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)), end, start)
+        else :
+            step = LiteralInteger(1)
+        return Slice(start, end, step)
+
+    def _print_PyccelArraySize(self, expr):
+        return '{}.shape[{}]'.format(expr.arg, expr.index)
```

Comment:
Do we need to handle the ordering here? @saniQQ ?

Reply:
the ordering does not affect the shape of an array